### ISSUE TITLE
Chapter 5: done() call should be outside borwser.visit(....)

### ISSUE THREAD

**Issue by SinOfWrath** [EthanRBrown/web-development-with-node-and-express#50]
on pg 46 chap 5.
My tests first 2 test cases failed (i.e hood-river and oregon-coast) because it couldn't find done(),
(i dont know why) but it worked fine after i moved it out of function browser.visit(..).

**Comment by EthanRBrown** [comment]
If you move `done()` oustide of `browser.visit()`, you are defeating the purpose of the test and getting falsely passing tests.

The point of the call to `done()` is to allow asynchronous code to be tested.  `browser.visit()` is asynchronous, so it returns immediately.  Placing `done()` _after_ (rather than _inside_ the callback) means that the test will pass successfully...but you're not actually testing anything.  Any errors will happen in the callback, but the test will have already passed.

In other words, by moving `done()` outside of the callback, you're only making it _look_ as if the test is passing: it's still failing, but your test is now lying to you!

Put the call to `done()` back inside the callback: the problem is elsewhere.  Are you making sure to start the server (`node meadowlark.js`) in another terminal before running the tests?

**Comment by SinOfWrath** [comment]
i've tried after restarting the server but i'm still getting error:

1) Cross - Page Tests requesting a group rate quote  from the hood river tour pageshould populate the referrer field:
     Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test.
  

  2) Cross - Page Tests requesting a group rate from the oregon coast tour page should populate the referrer field:
     Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test.

**Comment by EthanRBrown** [comment]
Are you using the official repo?  The Chapter 5 tests (including crosspage) are passing for me.  What OS are you on?

**Comment by SinOfWrath** [comment]
I'm not using the official repo, i'll try with that now, and OS X El Capitan.

**Comment by EthanRBrown** [comment]
Ah, I'm glad you're not on Windows.  It's a real pain to get the tests working on Windows.

The timeouts mean that either the headless browser (Zombie) is not successfully loading the page, or is unable to click the link on the page when it loads.

**Comment by SinOfWrath** [comment]
same error with the official repo, version of  zombie i'm using is: 4.2.1, 
and thanks for the quick response.

**Comment by EthanRBrown** [comment]
You closed the issue; did you figure it out?

**Comment by SinOfWrath** [comment]
Nope, was not able to figure it out.
<img width="836" alt="screen shot 2015-12-22 at 3 01 04 am" src="https://cloud.githubusercontent.com/assets/15728282/11941832/5561b3b2-a858-11e5-9d90-3aba649c72b3.png">
 "dependencies": {
    "express": "^4.13.3",
    "express3-handlebars": "^0.5.2"
  },
  "devDependencies": {
    "chai": "^3.4.1",
    "mocha": "^2.3.4",
    "zombie": "^4.2.1"
  }

**Comment by EthanRBrown** [comment]
The next step in debugging is to see what calls (if any) are being made to the Meadowlark app.  First, install a logger; Morgan is the most popular:

    npm install --save morgan

Then link into the middleware, first thing after you create the app:

  app.use(require('morgan')('dev'));

Then re-start the Meadowlark app (`node meadowlark.js`) and re-run the tests.  You should (hoepfully) see some activity in the Meadowlark app window.

**Comment by SinOfWrath** [comment]
I ran the test case 3 times as you can see above from a different terminal window, also last case somehow has 4 GET calls, visit() is working fine there's some problem with clickLink().
<img width="997" alt="screen shot 2015-12-22 at 11 34 31 am" src="https://cloud.githubusercontent.com/assets/15728282/11948973/1495457c-a8a0-11e5-96d2-c5856757c1b8.png">

**Comment by markzero** [comment]
The thing here is that you have to use exact version of zombie as specified in package.json, which is ~2.0.0-alpha31.

It won't work for me also otherwise. When I console.log the document.referrer it's blank, but using this version of zombie gives me non-empty referrer. Hope this helps.

**Comment by EthanRBrown** [comment]
Good to know, @markzero; sorry I dropped off this thread.  It's been a busy week, and this fell off my radar.  I'm glad you got it figured out!  I should update the code to work with the latest version of Zombie...so much to do and so little time :(

**Comment by morizotter** [comment]
Me too.
I updated zombie to latest - 4.2.1 and `done()` was called. But error still occured. The cause is referrer. If I wrote as below, the test succeeded.

```
assert(browser.field('referrer').value === 'referrer');
```

**Comment by EthanRBrown** [comment]
Thanks, for that @morizotter...if you want to fix the repo to use a more recent version of Zombie, I would accept that PR...and integrate the changes into the upcoming book revision.

**Comment by reybango** [comment]
Pull request seems to have been merged but @assaf [mentioned](https://github.com/assaf/zombie/pull/947#issuecomment-145929059) that since the CI tests aren't passing he wasn't going to include it in the release. Makes sense. 

I went in and manually updated Zombie (which I hated to do for obvious reasons) with [this fix](https://github.com/YoussefKababe/zombie/commit/252196402cadb73b23a3182479b5e9e983774cc4) and it worked perfectly.

**Comment by philreyes** [comment]
Chapter 5 crosspage-tests did not work at first. ("Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test.") I also found that the referrer field was empty in the console after running the mocha test. Here is how I fixed it:
1. Included the extra-code (several files have to be amended, described there) from the WDNE github repo chapters 5 and 6
2. put the jquery-2.0.0.min.js in vendor/js/
3. Un-installed zombie 4.x (I had installed the latest version at first)
4. installed zombie 3.1.1
5. followed reybango's advice and manually updated zombie's history.js file 
Thanks everyone for helping me out here!

**Comment by ChadKreutzer** [comment]
I fixed it by updating zombie to the latest version, then changing the lines in its history.js file per 	[Youssef Kababe's pull request](https://github.com/assaf/zombie/pull/947/commits/252196402cadb73b23a3182479b5e9e983774cc4). after that, I modified the tests-crosspage.js file to use the [browser.localhost](https://github.com/assaf/zombie#browser) and [browser.assert.input](https://github.com/assaf/zombie#assert). I'm using cloud 9 at work, so I had to do some finagling with the IP and the PORT, but it's finally working.

**Comment by reynoldssb** [comment]
package.json

"devDependencies": {
    "chai": "^3.5.0",
    "mocha": "~1.20.1",
    "zombie": "~2.0.0-alpha31"
  }

>npm update
>sudo npm install -g mocha@1.20.1

>mocha -u tdd -R spec qa/tests-crosspage.js 2>/dev/null


  Cross-Page Tests
    ✓ requesting a group rate quote from the hood river tour page should populate the hidden referrer field correctly (1475ms)
    ✓ requesting a group rate from the oregon coast tour page should populate the hidden referrer field correctly (1291ms)
    ✓ visiting the "request group rate" page dirctly should result in an empty value for the referrer field (650ms)


  3 passing (3s)

**Comment by pdxrod** [comment]
Thank you, @markzero. It's taken me days to get Mocha tests working for a simple visit(page). I always got "_Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test_", no matter how closely I followed the examples at stackoverflow.com and wherever else on the web I searched, no matter how much I littered the code with 
`setTimeout(done, 4000)` 
and 
`.then(done, done);`.
As you said: "The thing here is that you have to use exact version of zombie as specified in package.json, which is ~2.0.0-alpha31.", meaning you need to add the line
`"zombie": "~2.0.0-alpha31"`
to your package.json file and then run
`npm install`.

**Comment by pdxrod** [comment]
Correction. My test works once. Then subsequent calls say
` Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.`
unless you delete folder node_modules and run
`npm install' to completely reinstall all the node modules. My code:
`
process.env.NODE_ENV = 'test';
var assert = require('assert');
var Browser = require('zombie');

describe('contact page', function() {
  before(function() {
    this.browser = new Browser({ site: 'http://www.google.com' });
  });

  before(function(done) {
    setTimeout(done, 4000);
    this.browser.visit('/', done);
  });

  beforeEach(function(){
    this.timeout(4000);
  });

  it('should show search form', function(done) {
     setTimeout(done, 4000);
     var browser = this.browser;
     browser.pressButton('Google Search').then(function() {
       assert.ok(browser.success);
       assert.equal(browser.text('h1'), 'Google')
     }).then(done, done);
   });

});

`

**Comment by pdxrod** [comment]
Try this example

http://krasimirtsonev.com/blog/article/Testing-with-headless-browser-Zombiejs-Jasmine-TDD-NodeJS-example

It works more than once in a row!
